urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/net/contrapunctus/lzma/ConcurrentBufferInputStream.java

```diff
@@ -0,0 +1,119 @@
+// ConcurrentBufferInputStream.java -- read bytes from blocking queue
+// Copyright (c)2007 Christopher League <league@contrapunctus.net>
+
+// This is free software, but it comes with ABSOLUTELY NO WARRANTY.
+// GNU Lesser General Public License 2.1 or Common Public License 1.0
+
+package net.contrapunctus.lzma;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InterruptedIOException;
+import java.io.PrintStream;
+import java.util.concurrent.ArrayBlockingQueue;
+
+class ConcurrentBufferInputStream extends InputStream
+{
+    protected ArrayBlockingQueue<byte[]> q;
+    protected byte[]                     buf  = null;
+    protected int                        next = 0;
+    protected boolean                    eof  = false;
+    
+    private static final PrintStream     dbg  = System.err;
+    private static final boolean         DEBUG;
+    
+    static
+    {
+        String ds = null;
+        try
+        {
+            ds = System.getProperty("DEBUG_ConcurrentBuffer");
+        }
+        catch (final SecurityException e)
+        {
+        }
+        DEBUG = ds != null;
+    }
+    
+    private ConcurrentBufferInputStream(ArrayBlockingQueue<byte[]> q)
+    {
+        if (DEBUG)
+        {
+            dbg.printf("%s << %s%n", this, q);
+        }
+        this.q = q;
+        eof = false;
+    }
+    
+    static InputStream create(ArrayBlockingQueue<byte[]> q)
+    {
+        final InputStream in = new ConcurrentBufferInputStream(q);
+        return in;
+    }
+    
+    protected byte[] guarded_take() throws IOException
+    {
+        try
+        {
+            return q.take();
+        }
+        catch (final InterruptedException exn)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)